### DISCUSSION TITLE
[CI] Automating the evolution of the source code

### DISCUSSION THREAD

**Discussion by Guikingone** [symfony/symfony#46593] [Q&A]
Hi everyone 👋🏻 

# Context

I'm opening this discussion to open a debate (or at least, introduce the idea and see where it goes) about the evolution of the source code of Symfony and the way it can be handled to ease the maintenance of it, helping to respect the PHP "standards" and bug fixes, introducing the next PHP features and in general, helping everyone during the process of maintaining the source code.

So, why this discussion? First and to be fully honest, I'm not a huge contributor to the Symfony source code, I might encounter some issues, be tempted to solve them via a PR but most of the time, I don't do it. Why? Well, first because I'm not confident about how to solve it and secondly, because I tend to sometimes "overfix" a bug by improving the overall code and I almost always finish with a PR that is way bigger than expected so, well, to prevent this, I don't contribute 😄 

So, what this discussion is about? Well, I'm a huge fan of refactoring and improving the source code of an application / tool / library when I can be but as you probably know, doing this type of actions can be pretty boring, errors prone and most of the time, we can miss new features, errors fixes, improvements and so on (who's capable of saying from which PHP version the usage of `@` no longer silent fatal errors? We often focus ourselves on the major features and forget about the overall improvements 🙁 ).

# Idea

In the PHP ecosystem, many tools have been built to improve our daily tasks and helping us keep our projects up-to-date, one of these tools is [Rector](https://github.com/rectorphp/rector) that you probably already know (if not, the documentation is the place to start), the idea behind it is to automate the refactoring phase and help everyone introducing new standards, new features, new "best practices" and many more, of course, the tool is not perfect (and I'm one of its first critics when it comes to handling _semver_ and many more subjects).

So, how Rector can help Symfony to improve its source code?

Well, difficult question, simple answer: 

> By automating the maintenance of it, helping new contributors like old ones to improve the overall quality and more important, introducing PHP latest features when it's possible and useful.

That's for the theory, but in practice, how?

First, here's the context that I used to test Rector against the Symfony actual code: 

- Branch `6.2`
- PHP `>= 8.1`
- Configuration file `rector.php`:

```php
<?php

declare(strict_types=1); # I know that Symfony does not use `strict_types`, that just a convention on my side

use Rector\Config\RectorConfig;
use Rector\Core\ValueObject\PhpVersion;
use Rector\Set\ValueObject\SetList;

return static function (RectorConfig $rectorConfig): void {
    $rectorConfig->phpVersion(phpVersion: PhpVersion::PHP_81);

    $rectorConfig->autoloadPaths(autoloadPaths: [
        __DIR__ . '/vendor/autoload.php',
    ]);

    $rectorConfig->paths(paths: [
        __DIR__ . '/src/Symfony/Component/Dotenv',
    ]);

    $rectorConfig->skip(criteria: [
        __DIR__ . '/vendor',
    ]);

    $rectorConfig->sets(sets: [
        SetList::CODE_QUALITY,
        SetList::DEAD_CODE,
        SetList::EARLY_RETURN,
        SetList::PHP_81,
        SetList::UNWRAP_COMPAT,
        SetList::TYPE_DECLARATION,
        SetList::TYPE_DECLARATION_STRICT,
    ]);
};
```

_Some rules have been applied to gave a wider spectrum of what Rector can do, the most important here is `SetList::PHP_81` as Symfony requires it since `6.1`_

So, why this configuration file? First, I don't want to test it against `HttpFoundation` or `HttpKernel` which are "core Symfony component" and contains some "magic", so I've launched it against `Runtime`, `DotEnv`, `Semaphore`, `PasswordHasher` and the contracts to obtain a good visual of the actual state of the code, nothing more.

So, what's the output?

Here it is: https://gist.github.com/Guikingone/fda615efc34e6bef34ba40a94db04e6c, the "applied rules" can be found in my fork: https://github.com/Guikingone/symfony/commit/2dd4387f0113292d0b99ec551a383c5ca9e00e86.

As you may notice, the majority if the "recommandations" are against the return types and some usages, how Symfony handle final classes and the newly introduced `readonly` properties, so, in the end, to compare it against a more representative component, here's the exact same configuration applied to `Runtime`: https://github.com/Guikingone/symfony/commit/a5373377ebe4c2a9e2290f5f7fcb8a6b28b9ded6, again, the majority of the recommandations are formulated against the tests (which by default, could contains some magic, that could be excluded if necessary) and the "pretty new features" that Symfony may not support for various reasons (BC breaks, performances, code standards, etc).

As you may noticed, some "improvements" are open to debate, some can be further improved, again, that's just a "try and see" case.

**PS: In both cases, the tests are green locally.**

# Goal

So, you may ask: Why all this examples in the end?

First, I'm not a contributor of Rector and I don't speak for the team behind it, this discussion is really here to see how the actual source code of Symfony can be improved to support new features, bug fixes and overall improvements, nothing more, nothing less, this is not a poll about "**To rector or not**" or any kind of debate about "**how Rector can be tweaked to improve Symfony**" (it can but that's another subject 😄 ), I know (and I've followed the debates) that tools like Psalm has been introduced to Symfony and that the overall quality has improved since, that discussions continues to be open PR by PR about why, how and when introducing and/or using new rules triggered by Psalm and that in the end, each tool is here to correct a miss or improve the overall life of daily maintainers.

So, enough talks for today, again, I'm not talking for the rector team and this discussion is really here to debate and see if ideas introduced by Rector can be applied to Symfony (I'm pretty sure some recommandations can already be applied) and if Symfony can benefit from it.

Thanks again for the time and sorry for the _huge_ post, there's a lot of things that can be discussed so, time to do it if you wish to 🙂

**Comment by lyrixx** [comment]
Hi

So what is the final question?

---

FYI, we usually don't accept CS / refactoring of very old branch. 

> If it ain't broke, don't fix it

---

For what it worth, I prefer when the code is more readable, and simpler.

In the past we have accepted [many PR](https://github.com/symfony/symfony/pulls?q=is%3Apr+ea+extended+is%3Aclosed) from @[kalessil](https://github.com/kalessil)

so feel free to open some PR, since you already patch the code.

It's better to discuss the code, than some ideas :)

**Reply by Guikingone** [reply]
The final question is: Could the source code be improved thanks to Rector (not via the CI but by separate PR's) and PR submitted to discuss about the improvements? 🙂

**Reply by lyrixx** [reply]
Sure.

**Comment by stof** [comment]
Be careful. Adding return types is a BC break for child class. So any return type on existing classes can only be added in Symfony 7.0 (we added lots of them in 6.0 already, but some of them were explicitly delayed to give more time to the ecosystem to catch up, especially when the added return type would require PHP 8+ while downstream libraries still wanted to support PHP 7.4 with Symfony 5.4).

**Reply by Guikingone** [reply]
Yes, I noticed the fact that the public API cannot be changed outside of major versions, question is, could we add return types for tests (like `: void`?) or is it not a good practice? 

As tests are by "design" privates, maybe it could be improved? 

In the end, return types can be submitted later when the `7.X` branch is created, that's not mandatory for now 🙂

**Reply by stof** [reply]
IIRC, we decided not to add the `void` return type on all tests because they don't provide much value.
